### PR TITLE
Add new "api" project template that supports native AOT

### PR DESCRIPTION
**Pull request by DamianEdwards**
This adds a new "api" project template that represents an HTTP API project that uses minimal APIs to implement two endpoints: one that returns a list of todos, and another that returns an todo with a given id. The todos are randomly generated once at startup:

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Logging.AddConsole();
var app = builder.Build();

var sampleTodos = TodoGenerator.GenerateTodos().ToArray();

var todosApi = app.MapGroup("/todos");
todosApi.MapGet("/", () => sampleTodos);
todosApi.MapGet("/{id}", (int id) =>
    sampleTodos.FirstOrDefault(a => a.Id == id) is { } todo
        ? Results.Ok(todo)
        : Results.NotFound());

app.Run();
```

Unlike the existing "webapi" project template (which is not going away), this project template is focused squarely at new, cloud-focused API projects, uses minimal APIs only (no MVC option), and has a first-class option to enable support for publishing as native AOT: `dotnet new api -aot`. When the `-aot` flag is passed, additional code is emitted in the project to ensure the JSON serializer is configured appropriately to work in native AOT (i.e. a `JsonSerializerContext` is declared and hooked up to DI). When native AOT is enabled this template currently produces projects that emit build/publish warnings but of course the point is for those to be addressed by the all the currently in-progress work for ASP.NET Core native AOT:

```csharp
using System.Text.Json.Serialization;
using Company.ApiApplication1;

var builder = WebApplication.CreateBuilder(args);
builder.Logging.AddConsole();

builder.Services.ConfigureHttpJsonOptions(options =>
{
    options.SerializerOptions.AddContext<AppJsonSerializerContext>();
});

var app = builder.Build();

var sampleTodos = TodoGenerator.GenerateTodos().ToArray();

var todosApi = app.MapGroup("/todos");
todosApi.MapGet("/", () => sampleTodos);
todosApi.MapGet("/{id}", (int id) =>
    sampleTodos.FirstOrDefault(a => a.Id == id) is { } todo
        ? Results.Ok(todo)
        : Results.NotFound());

app.Run();

[JsonSerializable(typeof(Todo[]))]
internal partial class AppJsonSerializerContext : JsonSerializerContext
{

}
```

In line with the goals we have for native AOT and this style of project, there are metrics we intend to hit for projects created with this template regarding publish output file size, startup time, working set, and throughput. At this point that means this template configures the project to use Workstation GC (via a property in the project file), perform full trimming, and suppresses the publishing of IIS-related assets. At a later stage these properties will likely be automatically defaulted to these values in the Web SDK itself based on the value of the `<PublishAot>` property, at which point this template will be updated. When the [new `WebApplication.CreateSlimBuilder()` API is added](https://github.com/dotnet/aspnetcore/pull/46040), this template will be updated to use it too.

This template is currently configured to be hidden in Visual Studio. We'll change it to show up when we're ready to include the overall ASP.NET Core native AOT story in the .NET 8 preview messaging (currently targeting preview.2).

Addresses dotnet/aspnetcore#45526

### CODE REVIEW

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs
Comment (halter73) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs:41]: Should we put this in a separate AppJsonSerializerContext.cs file? Is it easy enough to conditionally include files to the template?

**Review by halter73 — APPROVED** [review]
I don't like the "api" name for this template. I don't think most ASP.NET Core developers will understand the distinction from the "webapi" project template. It's also a bit presumptuous 😆 

I know we have our own goals for having a "cloud focused" template which we can add "-aot" to, but I think it's better to continue using "webapi" and live with some unsupported combinations with "-aot". Maybe "-aot" could imply "-minimal".

Also, am I right in assuming `-aot` will work and `--publish-native-aot` will work, but `--aot` won't work? Can we support `--aot`, because `-aot` is super weird. I have a similar problem with `-minimal` working but `--minimal` not in the webapi template.

That said, we can wait for feedback and adjust.

File: src/ProjectTemplates/scripts/Test-Template.ps1
Comment (DamianEdwards) [reply src/ProjectTemplates/scripts/Test-Template.ps1:61]: TBC this script isn't part of actual template tests, it's only used to support local dev of templates and these additions are intended to help "disconnect" the templates from the various `Directory.Build.props` in parent dirs from the repo. So, having them apply for all the helper scripts is useful IMO.

**Comment by DamianEdwards** [comment]
> I don't like the "api" name for this template. I don't think most ASP.NET Core developers will understand the distinction from the "webapi" project template. It's also a bit presumptuous 😆

> I know we have our own goals for having a "cloud focused" template which we can add "-aot" to, but I think it's better to continue using "webapi" and live with some unsupported combinations with "-aot". Maybe "-aot" could imply "-minimal".

We've seen evidence that "web API" is a term that isn't widely common outside of the .NET ecosystem and the existing "webapi" template has its current defaults that are problematic to change. I very much like the simplicity of `dotnet new api -aot`.

> Also, am I right in assuming -aot will work and --publish-native-aot will work, but --aot won't work? Can we support --aot, because -aot is super weird. I have a similar problem with -minimal working but --minimal not in the webapi template.

This is the standard pattern we use for our project template options, i.e. a full `--` prefixed option name and in specific cases a `-` prefix shortened alias.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs:41]: I'm in two minds right now and have made all the logical arguments in my head both ways, but landed here aesthetically.  Interested to hear others' thoughts.

**Comment by davidfowl** [comment]
```C#
[JsonSerializable(typeof(string))]
[JsonSerializable(typeof(Album))]
[JsonSerializable(typeof(Album[]))]
internal partial class AppJsonSerializerContext : JsonSerializerContext
{

}
```

Can we just return NotFound without the JSON string?

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Album.cs
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Album.cs:18]: Logged #46087 to track the idea, thanks. We'll need to wait for dependency flow before we can make that change.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (eerhardt) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:29]: is there supposed to be a `#endif` for this?

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (eerhardt) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:30]: I think we can leave the primitive types off. The source generator takes care of the transitivity. It's kind of weird to call out `string`, but not have all the other types in Album.

Technically, we'd only need `Album[]`.

File: src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in
Comment (eerhardt) [inline src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in:4]: Is there a way to say that this template only supports `net8.0`+? I'm not familiar with the templating stuff, so if it is already handled, disregard.

At a minimum, I wouldn't expect `net7.0` + `-aot` to be supported.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs
Comment (eerhardt) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.Main.cs:41]: One part of me thinks: "Put each top level type in its own file", because that's the way I would do it. But the way you've proposed it here makes sense as well. Having this tiny little class in its own file seems going to far.

**Comment by eerhardt** [comment]
> This is the standard pattern we use for our project template options, i.e. a full -- prefixed option name and in specific cases a - prefix shortened alias.

If that is already the pattern we use for templates, we should follow it. But this is not a valid POSIX compliant design for command line args.

https://learn.microsoft.com/en-us/dotnet/standard/commandline/syntax#aliases

>> POSIX short forms typically have a single leading hyphen followed by a single character.

This is so you can group them together: `git clean -xdf` is setting 3 flags at once.

> Happy to just change it to return a 404 with no body.

+1 from me.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:29]: Known issue in templating: https://github.com/dotnet/templating/issues/4988

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:30]: Yeah this is one of those areas where what's correct/better could be debated. I'm going to remove the result that relies on string and if folks think it's better to just specify the array type here I'm happy to change it to that.

File: src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in:13]: Already opened: https://github.com/dotnet/sdk/issues/29896

File: src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/Api-CSharp.csproj.in:4]: That's already handled in the `template.json` file:

```json
  "Framework": {
      "type": "parameter",
      "description": "The target framework for the project.",
      "datatype": "choice",
      "choices": [
        {
          "choice": "net8.0",
          "description": "Target net8.0"
        }
      ],
      "replaces": "net8.0",
      "defaultValue": "net8.0"
    },
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (captainsafia) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:23]: Nit: `SingleOrDefault` seems more appropriate for resolving an element by a unique identifier.

File: src/ProjectTemplates/scripts/Run-Angular-Locally.ps1
Comment (captainsafia) [inline src/ProjectTemplates/scripts/Run-Angular-Locally.ps1:12]: Not relate to this PR but is this something we should be updated regularly?

**Review by captainsafia — APPROVED** [review]

File: src/ProjectTemplates/scripts/Run-Angular-Locally.ps1
Comment (DamianEdwards) [reply src/ProjectTemplates/scripts/Run-Angular-Locally.ps1:12]: Just each major release.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:23]: During a recent review of our API scaffolders we were advised that `SingleOrDefault` has extra overhead that isn't worth incurring in these scenarios so I adopted it here too.

File: src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs
Comment (eerhardt) [inline src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:25]: @captainsafia @davidfowl -  How is code like this going to allow our minimal API source generator to infer that we need to add `Todo` to the JSON Source Generation? Will we look through the delegate and find all places that call `Results.X`? We won't be able to tell by the Delegate's return type, because here that will just be `IResult`, and not `IResult<Todo>`

Comment (DamianEdwards) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:25]: I spoke to @davidfowl about this recently and IIRC there isn't a firm plan for the SG to do this as yet. For now, the app would be responsible for ensuring returned types are declared on a JSON serializer context that is configured in DI. We talked about analyzers and code-fixers that could make the process of keeping these up to date easier, and of course ideally the minimal API SG would do this automatically as you suggest, but as I understand it that capability will not be there initially.

Comment (davidfowl) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:25]: That's right. There's currently no plan to call the JSON source generator from the minimal API one. All we have are vague ideas. I assume the JSON source generator would need to be refactored to do this, but it would be good to attempt this once we have the baseline functionality checked in.

Maybe @brunolins16 can investigate once we're done with the core JSON work.

Comment (davidfowl) [reply src/ProjectTemplates/Web.ProjectTemplates/content/Api-CSharp/Program.cs:25]: If you're interested though, I wrote some ghetto logic for this here:

https://github.com/davidfowl/uController/blob/7c1e8f3db1c6a20738eff792386696ff89d4d7e7/src/uController.SourceGenerator/uControllerGenerator.cs#L366

It was for a different purpose though, identifying the metadata for open api.

**Comment by ShreyasJejurkar** [comment]
And also do we have separate issue where we keep track of warning that we get from ILlinker when publishing this project!? It will easier to track what warnings we have addressed and what needs to be!

And also not sure we can do this or not. Do we have template tests so that we can see how many warnings we get when publishing this project and asset it to some int value. So that we know we got 3 warnings, we need to address that. Once we addresses a issue, we would just chnage the test to 2 value. That we can be sure, we are getting expected number of warnings and can get notify if we get additional warnings and can see whats changed and also where in aspnetcore internal core or code from runtime as there are constant code changes in aot model scenarios (at runtime repo).

**Pull request merged**
